syntax = "proto3";

package chat;

message ChatConnectionRequest {
  string username = 1;
}

message ChatConnectionResponse {
  string ip = 1;
  int32 port = 2;
  string message = 3;
  repeated PendingMessage pending_messages = 4; // Add this line
}

message PendingMessage {
  string group_id = 1;
  string message = 2;
}

message ConnectToChatRequest {
  string username = 1;
  string chat_id = 2;
}

message ConnectToChatResponse {
  string ip = 1;
  int32 port = 2;
  string message = 3;
}

message ClientInfo {
  string ip = 1;
  int32 port = 2;
}

message MessageRequest {
  string message = 1;
  string recipient_username = 2;
}

message MessageResponse {
  string message = 1;
}

message GroupChatRequest {
  string username = 1;
  string group_id = 2;
  string message = 3;
}

message GroupChatResponse {
  string message = 1;
}

message Empty {}

message DiscoveryResponse {
  string message = 1;
}

message InsultResponse {
  string message = 1;
}

service ChatService {
  rpc ConnectToChat(ChatConnectionRequest) returns (ChatConnectionResponse);
  rpc ConnectToChatById(ConnectToChatRequest) returns (ConnectToChatResponse);
  rpc GetClientInfo(ChatConnectionRequest) returns (ClientInfo);
  rpc SendMessage(MessageRequest) returns (MessageResponse);
  rpc ReceiveMessage(MessageRequest) returns (MessageResponse);
  rpc SubscribeToGroupChat(GroupChatRequest) returns (GroupChatResponse);
  rpc SendMessageToGroupChat(GroupChatRequest) returns (GroupChatResponse);
  rpc SubscribeToPersistentGroupChat(GroupChatRequest) returns (GroupChatResponse);
  rpc SendMessageToPersistentGroupChat(GroupChatRequest) returns (GroupChatResponse);
  rpc DiscoverChats(Empty) returns (DiscoveryResponse);
  rpc InsultChannel(Empty) returns (InsultResponse);
}

